### PR TITLE
✨ Add bond hash policy

### PR DESCRIPTION
**Pull request by shibaPuppy**
https://github.com/openstack/ironic/blob/master/ironic/tests/unit/drivers/modules/network/json_samples/network_data.json#L31
Adds the bond hash policy feature added to Ironic.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Review by dtantsur — CHANGES_REQUESTED** [review]
/ok-to-test

https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/helpers/openstack.py#L679 seems to require bond_links actually? I'm not sure how the link you pasted proves it's optional.

**Comment by shibaPuppy** [comment]
@dtantsur 
cloud-init code copies bond_links.
so I want to use it as optional
with this loophole, we use the bond interface as a dummy interface.


```sh
links:
- id: bond0
  type: bond
  ethernet_mac_address: ff:ff:ff:ff:ff:ff
  bond_links: []
- id: bond1
  type: bond
  ethernet_mac_address: 11:22:33:44:55:66
  bond_links:
    - eth2
    - eth3
....
```

File: api/v1alpha5/metal3datatemplate_types.go [hunk 1]
```diff
@@ -202,6 +202,11 @@ type NetworkDataLinkBond struct {
 	// balance-rr, active-backup, balance-xor, broadcast, balance-tlb, balance-alb, 802.3ad
 	BondMode string `json:"bondMode"`
 
+	// +kubebuilder:validation:Enum="layer2";"layer3+4";"layer2+3"
+	// Selects the transmit hash policy used for port selection in balance-xor and 802.3ad modes
+	// +optional
+	BondHashPolicy string `json:"bond_xmit_hash_policy"`
```
Comment (dtantsur) [inline api/v1alpha5/metal3datatemplate_types.go:208]: The JSON name does not conform to the used standard and does not fully match the field name. Why not just `bondHashPolicy`?

File: api/v1beta1/metal3datatemplate_types.go [hunk 2]
```diff
@@ -224,6 +224,11 @@ type NetworkDataLinkBond struct {
 	// balance-rr, active-backup, balance-xor, broadcast, balance-tlb, balance-alb, 802.3ad
 	BondMode string `json:"bondMode"`
 
+	// +kubebuilder:validation:Enum="layer2";"layer3+4";"layer2+3"
+	// Selects the transmit hash policy used for port selection in balance-xor and 802.3ad modes
+	// +optional
+	BondHashPolicy string `json:"bond_xmit_hash_policy"`
```
Comment (dtantsur) [inline api/v1beta1/metal3datatemplate_types.go:230]: same

File: baremetal/metal3data_manager.go [hunk 3]
```diff
@@ -762,12 +762,13 @@ func renderNetworkLinks(networkLinks infrav1.NetworkDataLink, bmh *bmov1alpha1.B
...
-			"mtu":                  link.MTU,
-			"ethernet_mac_address": macAddress,
-			"bond_mode":            link.BondMode,
-			"bond_links":           link.BondLinks,
+			"type":                  "bond",
+			"id":                    link.Id,
+			"mtu":                   link.MTU,
+			"ethernet_mac_address":  macAddress,
+			"bond_mode":             link.BondMode,
+			"bond_xmit_hash_policy": link.BondHashPolicy,
```
Comment (dtantsur) [inline baremetal/metal3data_manager.go:770]: I cannot find this field in the network data schema. Where is it coming from?

**Comment by shibaPuppy** [comment]
@kashifest @Rozzii 
please review

File: api/v1alpha5/metal3datatemplate_types.go [hunk 4]
```diff
@@ -202,6 +202,11 @@ type NetworkDataLinkBond struct {
 	// balance-rr, active-backup, balance-xor, broadcast, balance-tlb, balance-alb, 802.3ad
 	BondMode string `json:"bondMode"`
 
+	// +kubebuilder:validation:Enum="layer2";"layer3+4";"layer2+3"
+	// Selects the transmit hash policy used for port selection in balance-xor and 802.3ad modes
+	// +optional
+	BondXmitHashPolicy string `json:"bond_xmit_hash_policy"`
```
[the remark this answers is not in the document]
Comment (shibaPuppy) [reply api/v1alpha5/metal3datatemplate_types.go]: @dtantsur 
I forgot it, change done.

**Review by dtantsur — APPROVED** [review]

**Comment by shibaPuppy** [comment]
@furkatgofurov7 
plz review

File: api/v1alpha5/metal3datatemplate_types.go [hunk 5]
```diff
@@ -202,6 +202,11 @@ type NetworkDataLinkBond struct {
 	// balance-rr, active-backup, balance-xor, broadcast, balance-tlb, balance-alb, 802.3ad
 	BondMode string `json:"bondMode"`
 
+	// +kubebuilder:validation:Enum="layer2";"layer3+4";"layer2+3"
+	// Selects the transmit hash policy used for port selection in balance-xor and 802.3ad modes
+	// +optional
+	BondXmitHashPolicy string `json:"bondXmitHashPolicy"`
```
Comment (furkatgofurov7) [inline api/v1alpha5/metal3datatemplate_types.go:208]: We shouldn't add the new field to v1alpha5 API IMO but only to v1beta1, the former should stay the same, only conversion should be changed.

/cc @kashifest opinions?

Comment (shibaPuppy) [reply api/v1alpha5/metal3datatemplate_types.go:208]: @furkatgofurov7 
remove the bondXmitHashPolicy in v1alpha5 ?

Comment (kashifest) [reply api/v1alpha5/metal3datatemplate_types.go:208]: @shibaPuppy can you update your PR then which adds the new field in v1b1 api only? v1a5 is deprecated and will be removed gradually #971

**Review by Rozzii — APPROVED** [review]
/lgtm
Looks good as far as I can tell.

**Comment by Rozzii** [comment]
PTAL @kashifest @mboukhalfa @lentzi90

**Comment by Rozzii** [comment]
@kashifest  seems like the test haven't been triggered

**Review by adilGhaffarDev — COMMENTED** [review]
/lgtm

**Pull request merged**